File: src/main/java/com/stockdownloader/backtest/OptionsBacktestEngine.java
package com.stockdownloader.backtest;

import com.stockdownloader.model.OptionType;
import com.stockdownloader.model.OptionsTrade;
import com.stockdownloader.model.PriceData;
import com.stockdownloader.strategy.OptionsStrategy;
import com.stockdownloader.util.BlackScholesCalculator;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Backtesting engine for options strategies. Simulates options trading against
 * historical underlying price data using Black-Scholes pricing for synthetic
 * option premiums. Supports dated expirations and strike price selection.
 *
 * The engine:
 * - Evaluates the strategy signal at each bar
 * - Uses Black-Scholes to price synthetic options at the target strike/expiry
 * - Tracks time decay (theta) and premium changes as the underlying moves
 * - Handles expiration: closes positions when DTE reaches zero
 * - Captures volume from the underlying for each trade entry
 * - Maintains an equity curve accounting for premium flow
 */
public final class OptionsBacktestEngine {

    private static final int CONTRACT_MULTIPLIER = 100;
    private static final BigDecimal DEFAULT_RISK_FREE_RATE = new BigDecimal("0.05");
    private static final int VOLATILITY_LOOKBACK = 20;

    private final BigDecimal initialCapital;
    private final BigDecimal commission;
    private final BigDecimal riskFreeRate;

    public OptionsBacktestEngine(BigDecimal initialCapital, BigDecimal commission) {
        this(initialCapital, commission, DEFAULT_RISK_FREE_RATE);
    }

    public OptionsBacktestEngine(BigDecimal initialCapital, BigDecimal commission,
                                 BigDecimal riskFreeRate) {
        this.initialCapital = Objects.requireNonNull(initialCapital);
        this.commission = Objects.requireNonNull(commission);
        this.riskFreeRate = Objects.requireNonNull(riskFreeRate);
    }

    public OptionsBacktestResult run(OptionsStrategy strategy, List<PriceData> data) {
        Objects.requireNonNull(strategy, "strategy must not be null");
        if (data == null || data.isEmpty()) {
            throw new IllegalArgumentException("data must not be null or empty");
        }

        var result = new OptionsBacktestResult(strategy.getName(), initialCapital);
        BigDecimal cash = initialCapital;
        OptionsTrade currentTrade = null;
        int tradeEntryBar = -1;
        BigDecimal tradeStrike = BigDecimal.ZERO;
        int tradeDTE = 0;
        List<BigDecimal> equityCurve = new ArrayList<>(data.size());

        result.setStartDate(data.getFirst().date());
        result.setEndDate(data.getLast().date());

        // Pre-compute close prices for volatility estimation
        BigDecimal[] closePrices = data.stream().map(PriceData::close).toArray(BigDecimal[]::new);

        for (int i = 0; i < data.size(); i++) {
            PriceData bar = data.get(i);
            BigDecimal spotPrice = bar.close();

            // Calculate current position value
            BigDecimal equity = cash;
            if (currentTrade != null && currentTrade.getStatus() == OptionsTrade.Status.OPEN) {
                int barsHeld = i - tradeEntryBar;
                int remainingDTE = Math.max(tradeDTE - barsHeld, 0);
                BigDecimal timeToExpiry = BigDecimal.valueOf(remainingDTE).divide(BigDecimal.valueOf(365), 10, RoundingMode.HALF_UP);
                BigDecimal vol = BlackScholesCalculator.estimateVolatility(closePrices, Math.min(i + 1, VOLATILITY_LOOKBACK));

                BigDecimal currentPremium = BlackScholesCalculator.price(
                        strategy.getOptionType(), spotPrice, tradeStrike,
                        timeToExpiry, riskFreeRate, vol);

                BigDecimal positionValue = currentPremium
                        .multiply(BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER));

                if (strategy.isShort()) {
                    // Short position: we received premium, owe the current value
                    equity = cash.subtract(positionValue).add(currentTrade.totalEntryCost());
                } else {
                    // Long position: value is the current premium
                    equity = cash.add(positionValue);
                }
            }
            equityCurve.add(equity);

            // Check expiration
            if (currentTrade != null && currentTrade.getStatus() == OptionsTrade.Status.OPEN) {
                int barsHeld = i - tradeEntryBar;
                if (barsHeld >= tradeDTE) {
                    // Option expired
                    BigDecimal intrinsic = BlackScholesCalculator.intrinsicValue(
                            strategy.getOptionType(), spotPrice, tradeStrike);

                    if (strategy.isShort()) {
                        currentTrade.expire(bar.date(), intrinsic);
                        cash = cash.subtract(intrinsic.multiply(
                                BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
                        cash = cash.subtract(commission);
                    } else {
                        currentTrade.expire(bar.date(), intrinsic);
                        cash = cash.add(intrinsic.multiply(
                                BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
                        cash = cash.subtract(commission);
                    }
                    result.addTrade(currentTrade);
                    currentTrade = null;
                    continue;
                }
            }

            // Evaluate strategy signal
            OptionsStrategy.Signal signal = strategy.evaluate(data, i);

            if (signal == OptionsStrategy.Signal.OPEN && currentTrade == null) {
                BigDecimal vol = BlackScholesCalculator.estimateVolatility(
                        closePrices, Math.min(i + 1, VOLATILITY_LOOKBACK));
                tradeStrike = strategy.getTargetStrike(spotPrice);
                tradeDTE = strategy.getTargetDaysToExpiry();
                BigDecimal timeToExpiry = BigDecimal.valueOf(tradeDTE)
                        .divide(BigDecimal.valueOf(365), 10, RoundingMode.HALF_UP);

                BigDecimal premium = BlackScholesCalculator.price(
                        strategy.getOptionType(), spotPrice, tradeStrike,
                        timeToExpiry, riskFreeRate, vol);

                if (premium.compareTo(BigDecimal.ZERO) <= 0) continue;

                // Determine number of contracts based on available capital
                int contracts;
                if (strategy.isShort()) {
                    // For short options, require margin (use underlying price as collateral)
                    BigDecimal marginPerContract = spotPrice.multiply(BigDecimal.valueOf(CONTRACT_MULTIPLIER));
                    contracts = cash.subtract(commission)
                            .divide(marginPerContract, 0, RoundingMode.DOWN)
                            .intValue();
                } else {
                    // For long options, cost is the premium
                    BigDecimal costPerContract = premium.multiply(BigDecimal.valueOf(CONTRACT_MULTIPLIER));
                    contracts = cash.subtract(commission)
                            .divide(costPerContract, 0, RoundingMode.DOWN)
                            .intValue();
                }

                contracts = Math.min(contracts, 10); // Cap at 10 contracts for risk management

                if (contracts > 0) {
                    OptionsTrade.Direction dir = strategy.isShort()
                            ? OptionsTrade.Direction.SELL : OptionsTrade.Direction.BUY;

                    currentTrade = new OptionsTrade(
                            strategy.getOptionType(), dir, tradeStrike,
                            bar.date(), bar.date(), premium, contracts, bar.volume());
                    tradeEntryBar = i;

                    if (strategy.isShort()) {
                        cash = cash.add(currentTrade.totalEntryCost());
                    } else {
                        cash = cash.subtract(currentTrade.totalEntryCost());
                    }
                    cash = cash.subtract(commission);
                }
            } else if (signal == OptionsStrategy.Signal.CLOSE
                    && currentTrade != null
                    && currentTrade.getStatus() == OptionsTrade.Status.OPEN) {
                int barsHeld = i - tradeEntryBar;
                int remainingDTE = Math.max(tradeDTE - barsHeld, 0);
                BigDecimal timeToExpiry = BigDecimal.valueOf(remainingDTE)
                        .divide(BigDecimal.valueOf(365), 10, RoundingMode.HALF_UP);
                BigDecimal vol = BlackScholesCalculator.estimateVolatility(
                        closePrices, Math.min(i + 1, VOLATILITY_LOOKBACK));

                BigDecimal exitPremium = BlackScholesCalculator.price(
                        strategy.getOptionType(), spotPrice, tradeStrike,
                        timeToExpiry, riskFreeRate, vol);

                if (strategy.isShort()) {
                    cash = cash.subtract(exitPremium.multiply(
                            BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
                } else {
                    cash = cash.add(exitPremium.multiply(
                            BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
                }
                cash = cash.subtract(commission);

                currentTrade.close(bar.date(), exitPremium);
                result.addTrade(currentTrade);
                currentTrade = null;
            }
        }

        // Force close any remaining open position at the last bar
        if (currentTrade != null && currentTrade.getStatus() == OptionsTrade.Status.OPEN) {
            PriceData lastBar = data.getLast();
            BigDecimal intrinsic = BlackScholesCalculator.intrinsicValue(
                    strategy.getOptionType(), lastBar.close(), tradeStrike);

            if (strategy.isShort()) {
                cash = cash.subtract(intrinsic.multiply(
                        BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
            } else {
                cash = cash.add(intrinsic.multiply(
                        BigDecimal.valueOf((long) currentTrade.getContracts() * CONTRACT_MULTIPLIER)));
            }
            cash = cash.subtract(commission);
            currentTrade.close(lastBar.date(), intrinsic);
            result.addTrade(currentTrade);
        }

        result.setFinalCapital(cash);
        result.setEquityCurve(equityCurve);
        return result;
    }
}

File: stockdownloader/backtest/options_backtest_engine.py
"""Options backtesting simulation engine using Black-Scholes pricing."""

import math
from stockdownloader.model.price_data import PriceData
from stockdownloader.model.option_type import OptionType
from stockdownloader.model.options_trade import OptionsTrade, OptionsDirection, OptionsTradeStatus
from stockdownloader.strategy.options_strategy import OptionsStrategy, OptionsSignal
from stockdownloader.backtest.options_backtest_result import OptionsBacktestResult
from stockdownloader.util import black_scholes


class OptionsBacktestEngine:
    CONTRACT_MULTIPLIER = 100

    def __init__(self, initial_capital: float = 100_000.0, commission_per_contract: float = 0.65,
                 option_type: OptionType = OptionType.CALL, direction: OptionsDirection = OptionsDirection.SELL,
                 risk_free_rate: float = 0.05):
        self.initial_capital = initial_capital
        self.commission_per_contract = commission_per_contract
        self.option_type = option_type
        self.direction = direction
        self.risk_free_rate = risk_free_rate

    def run(self, strategy: OptionsStrategy, data: list[PriceData]) -> OptionsBacktestResult:
        if not data:
            raise ValueError("data must not be empty")

        result = OptionsBacktestResult(strategy.get_name(), self.initial_capital)
        cash = self.initial_capital
        current_trade: OptionsTrade | None = None
        equity_curve: list[float] = []

        result.start_date = data[0].date
        result.end_date = data[-1].date

        vol_lookback = 20

        for i, bar in enumerate(data):
            signal = strategy.evaluate(data, i)

            equity = cash
            if current_trade and current_trade.status == OptionsTradeStatus.OPEN:
                close_prices = [data[j].close for j in range(max(0, i - vol_lookback), i + 1)]
                vol = black_scholes.estimate_volatility(close_prices, vol_lookback)
                dte = strategy.get_days_to_expiry()
                time_to_expiry = max(dte / 365.0, 1 / 365.0)
                current_premium = black_scholes.price(
                    self.option_type, bar.close, current_trade.strike,
                    time_to_expiry, self.risk_free_rate, vol
                )
                position_value = current_premium * current_trade.contracts * self.CONTRACT_MULTIPLIER
                if self.direction == OptionsDirection.SELL:
                    equity = cash + current_trade.total_entry_cost() - position_value
                else:
                    equity = cash + position_value
            equity_curve.append(equity)

            if signal == OptionsSignal.OPEN and current_trade is None:
                close_prices = [data[j].close for j in range(max(0, i - vol_lookback), i + 1)]
                vol = black_scholes.estimate_volatility(close_prices, vol_lookback)
                strike = strategy.get_strike_price(data, i)
                dte = strategy.get_days_to_expiry()
                time_to_expiry = dte / 365.0

                premium = black_scholes.price(
                    self.option_type, bar.close, strike,
                    time_to_expiry, self.risk_free_rate, vol
                )

                if premium > 0:
                    contracts = max(1, int(bar.close * 100 / (self.initial_capital * 0.1)))
                    contracts = min(contracts, 10)
                    commission = self.commission_per_contract * contracts

                    current_trade = OptionsTrade(
                        option_type=self.option_type,
                        direction=self.direction,
                        strike=strike,
                        expiration_date=bar.date,
                        entry_date=bar.date,
                        entry_premium=premium,
                        contracts=contracts,
                        entry_volume=bar.volume,
                    )

                    if self.direction == OptionsDirection.BUY:
                        cash -= premium * contracts * self.CONTRACT_MULTIPLIER + commission
                    else:
                        cash += premium * contracts * self.CONTRACT_MULTIPLIER - commission

            elif signal == OptionsSignal.CLOSE and current_trade and current_trade.status == OptionsTradeStatus.OPEN:
                close_prices = [data[j].close for j in range(max(0, i - vol_lookback), i + 1)]
                vol = black_scholes.estimate_volatility(close_prices, vol_lookback)
                time_to_expiry = max(strategy.get_days_to_expiry() / 365.0, 1 / 365.0)

                exit_premium = black_scholes.price(
                    self.option_type, bar.close, current_trade.strike,
                    time_to_expiry, self.risk_free_rate, vol
                )

                commission = self.commission_per_contract * current_trade.contracts
                current_trade.close(bar.date, exit_premium)

                if self.direction == OptionsDirection.BUY:
                    cash += exit_premium * current_trade.contracts * self.CONTRACT_MULTIPLIER - commission
                else:
                    cash -= exit_premium * current_trade.contracts * self.CONTRACT_MULTIPLIER + commission

                result.add_trade(current_trade)
                current_trade = None

        # Close any open position at end
        if current_trade and current_trade.status == OptionsTradeStatus.OPEN:
            last_bar = data[-1]
            close_prices = [d.close for d in data[-vol_lookback:]]
            vol = black_scholes.estimate_volatility(close_prices, vol_lookback)
            exit_premium = black_scholes.price(
                self.option_type, last_bar.close, current_trade.strike,
                1 / 365.0, self.risk_free_rate, vol
            )
            commission = self.commission_per_contract * current_trade.contracts
            current_trade.close(last_bar.date, exit_premium)

            if self.direction == OptionsDirection.BUY:
                cash += exit_premium * current_trade.contracts * self.CONTRACT_MULTIPLIER - commission
            else:
                cash -= exit_premium * current_trade.contracts * self.CONTRACT_MULTIPLIER + commission

            result.add_trade(current_trade)

        result.final_capital = cash
        result.equity_curve = equity_curve
        return result
